### PR TITLE
Switch to use Bootloader.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Follow up to #8547.

Move all migration operations out of the Server and into the Bootloader.

## How
* Add the `CONFIGS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION` and `JOBS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION`. These are env vars that will determine if the database is ready for an application to start.
* Add the `CONFIGS_DATABASE_INITIALIZATION_TIMEOUT_MS` and the `JOBS_DATABASE_INITIALIZATION_TIMEOUT_MS` env vars to determine how long an application should wait for the DB before giving up.
* Create the `MinimumFlywayMigrationVersionCheck` class. This class contains all the assertions to check if 1) a database is initialised. 2) a database meets the minimum migration version.
* Remove all set up operations from the ServerApp. Use `MinimumFlywayMigrationVersionCheck` operations instead.
* I also had to modify the `Databases` and `BaseDatabaseInstance` classes to support connecting to a database with timeouts. We would previously try forever.
* Add `Bootloader` to the relevant docker files and Kube files.

## Recommended reading order
1. `Configs.java` and `EnvConfigs.java` to see the new env vars.
2. `MinimumFlywayMigrationVersionCheck.java`, `Databases.java` and `BaseDatabaseInstance.java` to understand how we block and check the database is ready.
3. `ServerApp.java` to see how `MinimumFlywayMigrationVersionCheck.java` is used.

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-server/src/test/java/io/airbyte/server/ServerAppTest.java [hunk 1]
```diff
@@ -1,37 +0,0 @@
...
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-
-import io.airbyte.commons.version.AirbyteVersion;
-import org.junit.jupiter.api.Test;
-
-class ServerAppTest {
-
-  @Test
-  void testIsLegalUpgradePredicate() {
```
Comment (davinchia) [inline airbyte-server/src/test/java/io/airbyte/server/ServerAppTest.java:16]: these were moved to the bootloader test class in an eariler PR>

**Comment by davinchia** [comment]
Still working on getting the test 100% - the general structure is there so this is ready for a review!

File: airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java [hunk 2]
```diff
@@ -57,37 +57,36 @@ public BootloaderApp() {
...
+      createWorkspaceIfNoneExists(configRepository);
+
+      final JobPersistence jobPersistence = new DefaultJobPersistence(jobDatabase);
+      createDeploymentIfNoneExists(jobPersistence);
+      LOGGER.info("Set up job database and default deployment..");
+
+      final AirbyteVersion currAirbyteVersion = configs.getAirbyteVersion();
+      assertNonBreakingMigration(jobPersistence, currAirbyteVersion);
+
+      runFlywayMigration(configs, configDatabase, jobDatabase);
```
Comment (subodh1810) [inline airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:81]: As discussed running flyway migration should be the first step before we make any interaction with the database. We are creating a workspace, and a deployment above before executing the migration but as in my config migration, the DatabaseConfigPersistence will try to interact with `workspace` table which has not been created yet.

File: .env [hunk 3]
```diff
@@ -8,12 +8,14 @@ DATABASE_PORT=5432
 DATABASE_DB=airbyte
 # translate manually DATABASE_URL=jdbc:postgresql://${DATABASE_HOST}:${DATABASE_PORT/${DATABASE_DB} (do not include the username or password here)
 DATABASE_URL=jdbc:postgresql://db:5432/airbyte
+JOBS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION=0.29.15.001
```
Comment (subodh1810) [inline .env:11]: can you help me understand how these values would be updated in cloud?

**Review by subodh1810 — APPROVED** [review]
This makes sense! I just have 1 question and 1 blocking comment.

File: airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java (hunk 2 above)
Comment (davinchia) [reply airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:81]: good point!

File: .env (hunk 3 above)
Comment (davinchia) [reply .env:11]: We would have to specify these values separately in our Helm manifests when we upgrade to a newer version of OSS. So I'll have to do some work in the manifest when we update Cloud to the version of Airbyte that contains the bootloader.

Does that answer your question?

File: airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java (hunk 2 above)
Comment (davinchia) [reply airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:74]: @subodh1810 how is this?

**Comment by davinchia** [comment]
Ran this twice locally to confirm this works.

**Pull request merged**